### PR TITLE
AL09 (Self-alias): Quote alias when case changed

### LINKED ISSUE #5636: AL09 - Case sensitivity with unquoted identifiers: TSQL
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

When using sql fluff fix it rewrites our SQL queries and removes "self aliasing" even though the casing is different and even though this is nescessary.

### Expected Behaviour

I do not want "ID" to be interpreted as equal to "id" or "Id".

### Observed Behaviour

The following is changed by sqlfluff fix:
```
SELECT
	id AS  ID,
	date AS Date,
	description AS Description,
FROM
[...]
```
into:
```
SELECT
	id,
	date,
	description,
FROM
[...]
```

### How to reproduce

Use the above SQL query and run sqlfluff lint or fix on it

### Dialect

tsql

### Version

  - repo: https://github.com/sqlfluff/sqlfluff
    rev: 3.0.0a5
    hooks:
      - id: sqlfluff-fix
      - id: sqlfluff-lint

### Configuration

[tool.sqlfluff.core]
large_file_skip_byte_limit = 100000
dialect = "tsql"
ignore = "templating"

[tool.sqlfluff.indentation]
tab_space_size = "4"
indent_unit = "tab"

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### LINKED ISSUE #6019: AL09 - Case sensitivity with unquoted identifiers: Clickhouse
### Search before asking

- [X] I searched the [issues](https://github.com/sqlfluff/sqlfluff/issues) and found no similar issues.

### What Happened

The fix command unexpectedly deletes aliases when ignoring capitalization rules for identifiers.

Submitted query:
SELECT
    myidcol AS col_id,
    my_expected_drop_alias AS my_expected_drop_alias,
    My_Wrong_Case_Col as my_wrong_case_col,-- noqa: CP02
    my_good_col,
FROM
    file('myfile.parquet', 'Parquet')

Reformatted query:
SELECT
    myidcol AS col_id,
    my_expected_drop_alias,
    My_Wrong_Case_Col,-- noqa: CP02
    my_good_col,
FROM
    file('myfile.parquet', 'Parquet')

### Expected Behaviour

Expected result:
SELECT
    myidcol AS col_id,
    my_expected_drop_alias,
    My_Wrong_Case_Col as my_wrong_case_col,-- noqa: CP02
    my_good_col,
FROM
    file('myfile.parquet', 'Parquet')

### Observed Behaviour

When AL09 is used, but CP02 is ignored, AL09 is incorrectly triggered.

### How to reproduce

Example queries provided

### Dialect

Clickhouse

### Version

sqlfluff, version 3.1.0

### Configuration

Default config + inline noqa flag.

### Are you willing to work on and submit a PR to address the issue?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/sqlfluff/sqlfluff/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by aayushr7**
### Brief summary of the change made

This change changes the fix scenario when the self-alias is done for case change. For case change, users often do not quote the alias, example: `id AS ID`. The existing logic will fix this to `id`. However, users expect this as case change.

If AL09 passes the rule, this can conflict with capitalisation policy rules. For example: if `capitalisation.identifiers` is set to `lower`, then the final fix would be `id AS id` which is not expected.

Therefore, the alias should not be unquoted when there is case change, rather quoted. 
Correct fix: `id AS "ID"`. For BigQuery: ```id as `ID`.```
However, this fix has conflicts with RF06 which considers quotes unnecessary. Therefore, for such cases fix is not applied yet.

- fixes #5636
- fixes #6019

### Are there any other side effects of this change that we should be aware of?
None

### Pull Request checklist
- [X] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

Thanks @keraion for the initial PR.

### CODE REVIEW

**Comment by keraion** [comment]
This seems to introduce a number incompatibilities with other rules.

```sh
echo 'SELECT 
    COL1 as col1,
    COL2 as "col2"
FROM TAB;
' | sqlfluff fix --dialect ansi -
WARNING    Fixes for LT01 not applied, as it would result in an unparsable file. Please report this as a bug with a minimal query which demonstrates this warning.                                                                                                                                                        
WARNING    Fixes for LT12 not applied, as it would result in an unparsable file. Please report this as a bug with a minimal query which demonstrates this warning.                                                                                                                                                        
WARNING    One fix for AL09 not applied, it would re-cause the same error.                                                                                                                                                                                                                                                
SELECT                                                                                                                                                                                                                                                                                                                    
    COL1,
    COL2
FROM TAB;
```

Some additional conflicts when removing CP02
```sh
echo 'SELECT 
    COL1 as col1,
    COL2 as "col2"
FROM TAB;
' | sqlfluff fix --dialect ansi --exclude-rules CP02 -
WARNING    Fixes for LT01 not applied, as it would result in an unparsable file. Please report this as a bug with a minimal query which demonstrates this warning.                                                                                                                                                        
WARNING    Fixes for LT12 not applied, as it would result in an unparsable file. Please report this as a bug with a minimal query which demonstrates this warning.                                                                                                                                                        
WARNING    One fix for AL09 not applied, it would re-cause the same error.                                                                                                                                                                                                                                                
WARNING    One fix for RF06 not applied, it would re-cause the same error.                                                                                                                                                                                                                                                
WARNING    One fix for RF06 not applied, it would re-cause the same error.                                                                                                                                                                                                                                                
SELECT                                                                                                                                                                                                                                                                                                                    
    COL1 AS "col1",
    COL2 AS "col2"
FROM TAB;
```

**Comment by aayushr7** [comment]
Good spot @keraion. There is a conflict with `RF06` which considers quoted identifiers unnecessary and removes them during fix. However, I think they should not be considered unnecessary since they are changing the case of the column. Let me create an issue for that to investigate.

I have removed `fix` from `AL09` for such cases for now.

File: test/fixtures/rules/std_rule_cases/AL09.yml [hunk 1]
```diff
@@ -20,79 +20,79 @@ test_pass_no_self_alias_function:
...
 
-test_fail_self_alias:
-    fail_str:
+test_pass_self_alias_sensitivity_change:
+    # Renaming the column name from case-sensitive to case-insensitive
+    pass_str:
         select
-            col_a as col_a,
-            col_b as new_col_b
+            "col_a" as col_a,
```
Comment (WittierDinosaur) [inline test/fixtures/rules/std_rule_cases/AL09.yml:27]: I'm not sure we'd want these to pass - they are the same content, just with quotes. Thoughts?

Comment (aayushr7) [reply test/fixtures/rules/std_rule_cases/AL09.yml:27]: Actually, `"col_a"` is case sensitive, which results in the final column name to be all lower case. However, `col_a` without quotes results in a case-insensitive column name. 

Therefore, the alias here is changing the case sensitivity of the column names. 

Here is an example:

Screenshot 2024-08-11 at 22 34 07

Comment (aayushr7) [reply test/fixtures/rules/std_rule_cases/AL09.yml:27]: In this example, a CTAS command is used where the case-sensitivity of the column is changed. 

Screenshot 2024-08-11 at 22 43 45

Comment (aayushr7) [reply test/fixtures/rules/std_rule_cases/AL09.yml:27]: Hi @WittierDinosaur , yes! I had been testing with incorrect dialects all along. This case is more relevant to dialects like Oracle, Snowflake

**Comment by alanmcruickshank** [comment]
Closing this as superseded by #6333.

**Pull request closed without merging**